test/e2e/podtimingcontroller/podtimingcontroller.go

```diff
@@ -0,0 +1,90 @@
+//go:build e2e
+// +build e2e
+
+package podtimingcontroller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+func SetupWithManager(mgr ctrl.Manager, log logr.Logger, artifactDir string) error {
+	r := &podTimingReconciler{
+		client: mgr.GetClient(),
+		podEnterPhaseDurationSeconds: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "e2e_pod_enter_phase_seconds",
+		}, []string{"phase", "namespace", "name", "uid"}),
```

Comment:
thinking about this some more.... will this be sufficient to correlate similar pods over time? We actually don't care about the discrete identities of pods (e.g. UID) and each label will explode the number of series. The names will also be in most cases dynamically generated. Will that prevent them from being compared other than through like regex name comparisons?

I was thinking something like the following labels:
- the hosted cluster ID (infra ID should be a sufficient proxy for a hosted cluster's unique identity)
- pod's logical component (e.g. its deployment name?). The challenge for the logical component name is how to get it. Not sure we have a consistent labelling strategy here, but maybe we should introduce one.
- phase

Something else I just thought of regarding the counter... if the pod has a stable name/identity and is recreated, the counter could decrease, violating the metric type invariant. So maybe uid is required after all?

Reply:
Good points... we do have to start somewhere, I'm happy to just keep going with what you already have so we can start getting all the machinery in place and establish a foundation for further iteration.